### PR TITLE
Webhooks: Introduce `[webhooks][timeout]` configuration option to control the maximum timeout of Webhook notifications.

### LINKED ISSUE #73657: Alerting: unable to configure webhook notification timeout
### What happened?

We have a contact point configured with a webhook that tries to send the alerts to an external service that processes them.

Alerts triggered by the internal Grafana manager as multidimensional alert rules.

We usually have situations where the number of alerts are quite high and the process of the alerts can elapse more than 30s.

The following issue #58687 describes the same case, but it was closed as it derived into screenshot processes, not resolving the original issue.

So:
- Is there a way to increase the notification timeout?
- Is there a reason why the timeout is hardcoded to 30s?

### What did you expect to happen?

The notification answers OK with custom timeout on contact points configuration

### Did this work before?

No

### How do we reproduce it?

1. Up a simple http listener that answers after 30s
2. Create an alert rule that fires an alert
3. Configure contact point to use the http listener
4. Create a notifiaction rule that matches the rules
5. Wait until the alert is fired and check the health of the contact point/notification policy

### Is the bug inside a dashboard panel?

_No response_

### Environment (with versions)?

Grafana: 10.0.3

### Grafana platform?

Docker

### Datasource(s)?

_No response_

### PR DESCRIPTION
**Pull request by gotjosh**
Introduces a new configuration section and option that allows us to control the maximum timeout for Grafana's webhook sub-system.
This sub-system is leveraged by alerting to deliver notification to many of the integrations we support (e.g. Line, Telegram, Pagerduty - to name a few).

It's important to note two things:

1. When using Unified Alerting and Grafana is running in high-availability mode, the webhooks timeout must be less than `ha_peer_timeout` as we expect notifications to be delivered *before* the replica sending it fails over (via the peer timeout). To that avail, we've introduced a check that warns the operator as grafana is starting up.

2. This also affects legacy alerting, but we've reached EOL with it and I'd rather have it stay out of the documentation.

Fixes https://github.com/grafana/grafana/issues/73657

### CODE REVIEW

File: pkg/services/notifications/notifications.go [hunk 1]
```diff
@@ -241,20 +262,20 @@ func (ns *NotificationService) ValidateResetPasswordCode(ctx context.Context, qu
 		return nil, ErrInvalidEmailCode
 	}
 
-	user, err := userByLogin(ctx, login)
```
Comment (gotjosh) [inline pkg/services/notifications/notifications.go:244]: drive-by warning fix - package showing variable.

File: pkg/services/notifications/webhook.go [hunk 2]
```diff
@@ -32,21 +29,6 @@ type WebhookClient interface {
...
-		Renegotiation: tls.RenegotiateFreelyAsClient,
-	},
-	Proxy: http.ProxyFromEnvironment,
-	Dial: (&net.Dialer{
-		Timeout: 30 * time.Second,
-	}).Dial,
-	TLSHandshakeTimeout: 5 * time.Second,
-}
-var netClient WebhookClient = &http.Client{
-	Timeout:   time.Second * 30,
```
Comment (gotjosh) [inline pkg/services/notifications/webhook.go:46]: Moved as is to pkg/services/notifications/notifications.go, with the exception of this line which I used the setting read from the configuration file.

**Review by grobinson-grafana — DISMISSED** [review]
LGTM!

I have two small questions that are non-blocking:

1. I was expecting a PR for prometheus/alertmanager that we would then add to Grafana. Is there a reason we don't want to upstream this change? It seems quite useful?

2. The other question I had is what is the reasoning for doing this at the configuration-file level instead of at the contact point level? I can see different webhooks wanting different timeouts, and users in Grafana Cloud won't be able to change the timeout themselves without asking customer support? Is this something that we might want to reconsider?

File: pkg/services/notifications/notifications.go [hunk 3]
```diff
@@ -39,7 +43,23 @@ var tmplSignUpStarted = "signup_started"
 var tmplWelcomeOnSignUp = "welcome_on_signup"
 
 func ProvideService(bus bus.Bus, cfg *setting.Cfg, mailer Mailer, store TempUserStore) (*NotificationService, error) {
+	netTransport := &http.Transport{
+		TLSClientConfig: &tls.Config{
+			Renegotiation: tls.RenegotiateFreelyAsClient,
+		},
+		Proxy: http.ProxyFromEnvironment,
+		Dial: (&net.Dialer{
+			Timeout: 30 * time.Second,
```
Comment (alexweav) [inline pkg/services/notifications/notifications.go:52]: What is the meaning of the hardcoded 30s dial timeout (and 5s TLS handshake timeout) in relation to the `http.Client` timeout?

Are they additive, meaning in some cases requests could take 35s longer than the configured timeout? Or does the client timeout encompass the whole request?

If the `http.Client` timeout overrides all, which I **suspect** it does - could we consider relying on that purely and dropping the other timeouts?

Comment (gotjosh) [reply pkg/services/notifications/notifications.go:52]:
```
// Timeout specifies a time limit for requests made by this
// Client. The timeout includes connection time, any
// redirects, and reading the response body. The timer remains
// running after Get, Head, Post, or Do return and will
// interrupt reading of the Response.Body.
```

> What is the meaning of the hardcoded 30s dial timeout (and 5s TLS handshake timeout) in relation to the http.Client timeout?

The dial is the timeout for establishing the TCP connection whereas the timeout on the HTTP client is for the HTTP connection. 

I cite the documentation on the `Timeout` field as it illustrates that HTTP client connection does encompass everthing. Indeed the `http.Client` timeout means a maximum of 30s for _everything_ but I wouldn't drop the others as they mean something in their own merit e.g. a TCP connection establishing taking 25s, then 3s for the HTTP call would succeed.

To follow your suggestion, I would need to set all Timeouts to 0s effectively removing the timeouts for those other operations. I briefly paused and thought about that as I putting this PR together but chose not to for two reasons:

1. Alerting is probably the biggest user of the notifications service yes, but I couldn't quickly tell if it's the _only_ user. There's a lot of alerting integrations this code depends to and I'm not sure what kind of side-effects could that code change bring.

2. The problem that I'm trying to solve is very simplistic, more time for HTTP webhooks to succeed - making the suggested change yields no direct benefit (IMO) to that.

If you feel strongly about this - I'm happy to open a separate PR with the change but I would rather not do it under the assumption that this hasn't been a problem.

**Comment by gotjosh** [comment]
> I was expecting a PR for prometheus/alertmanager that we would then add to Grafana. Is there a reason we don't want to upstream this change? It seems quite useful?

Do you have a citation on its usefulness or is this a gut feeling? I'd need to investigate what that requires in the AM. But to answer your question - I didn't think about upstreaming this change therefore I have no reason besides the common "I have other priorities". 

> The other question I had is what is the reasoning for doing this at the configuration-file level instead of at the contact point level? I can see different webhooks wanting different timeouts, and users in Grafana Cloud won't be able to change the timeout themselves without asking customer support? Is this something that we might want to reconsider?

Allowing users (as opposed to grafana server administrators) to control the timeouts makes it harder to control SLI/SLOs on notification delivery. If we have a target of delivering all webhooks under `30s` and suddenly we're allowing every single user to set arbitrary timeouts this makes it harder to predict how long do we expect outgoing connections to be alive.

I thought about this too as I looked at how context is handled in notifiers but ultimately decided against it due to unwarrant complexity. I think that with enough support is something that we could consider but for now, I'd say this is good enough.

File: pkg/services/notifications/notifications.go (hunk 3 above)
Comment (alexweav) [reply pkg/services/notifications/notifications.go:52]: Nah, don't worry about it, just felt the duplicated 30 seemed unnecessary. It's probably not worth removing now.

Comment (grobinson-grafana) [reply pkg/services/notifications/notifications.go:52]: I did consider asking the same question Alex, but I saw it was just copied code. I am interested in understanding if there was a reason for reducing the TLS handshake timeout from a default of 10 seconds to 5 seconds. The TCP connection handshake (dialer) timeout is the same as the default, which is 30 seconds.

**Comment by grobinson-grafana** [comment]
I'm thinking about it from a number of different perspectives:

1. If we do this work upstream then we can use it in both Grafana and Grafana Cloud, and avoid the need to do the work twice if/when another customer wants this for Grafana Cloud.
2. If we make it a configuration option for Grafana server administrators, what does that mean for webhook contact points in the unified AM? How do we migrate this feature over?
3. I wasn't thinking about SLOs, is that from the perspective of Grafana Cloud? The picture I had painted was that there was a maximum timeout that could be set, but contact points would be allowed to choose a timeout within the range of [0, max_timeout]. A naunce here is the server wide timeout is larger than a route's `group_interval` then it can disrupt subsequent flushes and cause additional races with `peer_timeout`.

**Comment by grobinson-grafana** [comment]
I just dismissed the approval as I'm having issues getting 30 second timeouts to work.

I'm using the default configuration:

```
[webhooks]
# The timeout for delivering webhook notifications. Most of the alerting integrations use webhooks to deliver notifications.
# If you're running a Highly Available setup and use Unified Alerting this should be less than [unified_alerting][ha_peer_timeout] to avoid duplicate notifications.
# The timeout string is a possibly signed sequence of decimal numbers, followed by a unit suffix (ms, s, m, h, d), e.g. 30s or 1m
timeout = 30s
```

and I have a webhook server that waits 45 seconds for all HTTP requests. I can see the aggregation group being flushed and the webhook being sent:

```
DEBUG[10-21|10:55:00] flushing                                 logger=ngalert.notifier.alertmanager LOG15_ERROR= component=alertmanager orgID=1 component=dispatcher aggrGroup="{}:{alertname=\"Test\", grafana_folder=\"Test Folder\"}" alerts=[Test[f2db7a5][active]]
DEBUG[10-21|10:55:00] Sending webhook                          logger=notifications url=http://127.0.0.1:8080 http method=POST
```

However, the timeout occurs after 15 seconds, not 30 seconds:

```
ERROR[10-21|10:55:15] Notify for alerts failed                 logger=ngalert.notifier.alertmanager LOG15_ERROR= component=alertmanager orgID=1 component=dispatcher num_alerts=1 err="webhook/webhook[0]: notify retry canceled due to unrecoverable error after 1 attempts: Post \"http://127.0.0.1:8080\": context deadline exceeded"
```

I added a log line in `ProvideService` and I can see that the timeout used in the HTTP client is 30 seconds:

```
[INFO ] [10-21|10:58:22]                                          logger=notifications cfg.Webhooks.Timeout=30s
```
